### PR TITLE
[FLINK-24193][tests] Add ClassLoaderExtension 

### PR DESCRIPTION
**Pull request by zentol**
At various points in Flink we use the ServiceLoader mechanism to load implementations, for example filesystems or reporters.
It is thus only natural that we also have some test service implementations which are used in specific tests.

However, these test implementations are intended to only be used in very specific tests, but are currently but on the classpath for all tests in that module (+all that depend on the test-jar of such a module). This causes confusion (e.g., suddenly there are 5 reporter factories available) or logging noise (e.g., custom netty handlers being loaded by each MiniCluster).

This PR mplements a junit extension that runs the test with a customized classloader, which also has access to a temporary directory containing generated service entries.

The RestExternalHandlersITCase was migrated to use this extension, as this was the issue that prompted the whole effort.

### CODE REVIEW

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Function;
+
+/** JUnit extension to customize the classloader that a test is run with. */
+public class ClassLoaderExtension implements BeforeAllCallback, AfterAllCallback {
```
Comment (AHeise) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java:38]: `ContextClassLoaderExtension`?

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    private ClassLoader originalClassLoader;
+    private URLClassLoader temporaryClassLoader;
+
+    private ClassLoaderExtension(
+            Function<TemporaryFolder, URLClassLoader> temporaryClassLoaderFactory) {
+        this.temporaryClassLoaderFactory = temporaryClassLoaderFactory;
+    }
+
+    @Override
+    public void beforeAll(ExtensionContext context) throws Exception {
```
Comment (AHeise) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java:52]: We talked about that offline, but now this extension cannot be used per test, right?
What happens if we also implement `before/afterEach` would that allow us to use the extension both as a ClassRule and an instance Rule?

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    /**
+     * Creates a new {@link ClassLoaderExtension} that puts additional service entries on the
+     * classpath.
+     *
+     * @param entries service entries that should be added to the classpath
+     * @return ClassLoaderExtension
+     */
+    public static ClassLoaderExtension withServiceEntries(ServiceEntry... entries) {
```
Comment (AHeise) [inline flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java:73]: If this was production code, I'd push for some builder pattern etc. I wouldn't expose `ServiceEntry` directly. I'd be fine to leave as is (for now). Some ideas:

On instance method with RegisterExtension.
```
new ClassLoaderExtension().withServiceEntry(serviceClass, serviceImplementations).withServiceEntry(...)
```

```
ClassLoaderExtension.buidler().addServiceEntry(...).addServiceEntry(...).build()
```

Or more declaratively
```
@ExtendWith(ClassLoaderExtension.class)
class ServiceTest {
  @Test
  @WithServiceEntry(clazz, implementations = [])
  void testService() {
  }
}
```

**Review by AHeise — COMMENTED** [review]
Very good idea and clean approach. I added a few ideas on the usability aspect.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java (hunk 2 above)
Comment (zentol) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java:52]: I suppose junit would call both versions resulting in odd behavior.

I think the solution in the end will be a wrapper around such an extension that calls `beforeAll` in `beforeEach`.

Something like:
```

    private static class ForEachExtensionWrapper<T extends BeforeAllCallback & AfterAllCallback>
            implements BeforeEachCallback, AfterEachCallback {

        private final T backingExtension;

        private ForEachExtensionWrapper(T backingExtension) {
            this.backingExtension = backingExtension;
        }

        @Override
        public void beforeEach(ExtensionContext context) throws Exception {
            backingExtension.beforeAll(context);
        }

        @Override
        public void afterEach(ExtensionContext context) throws Exception {
            backingExtension.afterAll(context);
        }
    }
```

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java (hunk 3 above)
Comment (zentol) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java:73]: Yes in the future we'll need a builder pattern of sorts. If it doesn't take too long Ill add it in this PR.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java (hunk 2 above)
Comment (ruanhang1993) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java]: We are recently working at changing tests from Junit 4 to Junit 5. And the same problem occurs when the extension is used both as a ClassRule and an instance Rule.
I prefer to provide a `EachExtensionWrapper`(for beforeAll, afterAll) and a `AllExtensionWrapper`(for beforeEach, AfterEach) in the module `flink-test-utils-junit` like this to solve it.

Comment (zentol) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/ClassLoaderExtension.java]: Or possible something even more generic, where the extensions themselves don't even implement beforeEach/All, with those details completely being handled by the wrapper.

**Review by AHeise — APPROVED** [review]
LGTM % commit structure but I assume you squash most of it.

One more comment: you can remove most `public`s in your JUnit5 test.

**Pull request merged**